Allow user to purge old history entries

schlimmchen opened the pull request:
the database maintenance widget is complemented with a section that allows the user to clean up old history entries. the age of history entries to be removed is configurable. respective entries are searched for and removed with the click of a button.  message boxes inform about whether or not and if so how many entries were removed from the database.

Fixes #7550.

This implementation leans on the implementation in KeePass2. I was going to add a line to the database settings general tab alongside the "Max. history items" and "Max. history size" settings. That would either require a custom, i.e., incompatible, metadata datum, or would use the existing metadata datum in an incompatible way (KeePassXC would truncate history according to this setting at any chance it gets while the respective entries are only removed by KeePass2 by the click of a button).

## Screenshots
![image](https://user-images.githubusercontent.com/3578416/161442473-5660c783-784c-4cb4-83d3-f8da0848c4ed.png)
![image](https://user-images.githubusercontent.com/3578416/161442487-bd713566-e9c3-40b4-9e15-37e915fa29fa.png)
![image](https://user-images.githubusercontent.com/3578416/161442513-b9755b19-8167-4d26-9359-97cac95498ff.png)


## Testing strategy
Manual


## Type of change
- ✅ New feature (change that adds functionality)

droidmonkey commented:
Days is rather unapproachable, also we always want to leave at least the last history item unless the user specifically wants it all cleared out.  I would rather like something that gives you a choice:

- Remove all but latest history from all entries
- Remove all history from all entries
- Remove entry history older than [choice]

With choice being 6 months, 1 year, and 2+ years, perhaps with a type in option.

schlimmchen commented:
The user can clear out all history elements by setting the choice to zero. But I agree that this is not very intuitive or explicit.

I like the idea of removing all but the latest history entry!

I also understand that days is not a very useful unit to choose an age in. I don't like to mix the units. And I also don't think that we should limit the user to only a couple of pre-defined choices. There will be issues opened "meh, but I want this and that option" (see #7369). You suggest having one of the choices "in option", but I don't understand how that will work (user interface). How about we change the unit from days to months (and save it as days in the DB metadata)?.

The three modes of cleanup will be radio buttons, right?

schlimmchen commented:
Proposal:

![image](https://user-images.githubusercontent.com/3578416/161473801-b6f1656e-33c0-4b01-b0d8-aa5009de6542.png)

michaelk83 commented:
Days are still useful for small numbers (up to 90 or so?), and I'm sure some users may want to leave only the last 7 days, the last 10 days, etc. You can leave the unit as a choice, similar to https://github.com/keepassxreboot/keepassxc/issues/4896#issuecomment-648529307 .

Btw, password expiry and this have very similar UI requirements. Maybe worth implementing a common "time period" component?

schlimmchen commented:
> Maybe worth implementing a common "time period" component?

Probably I will agree with you. It makes this implementation more complex and I will not promise that I can handle all of it, but let's see.

Can you please elaborate a little? The UI Designs (.ui files) will still define their own copy of a spin box for the amount and a combo box for the unit, but there shall be a common class that "mounts" onto them for reading and writing? Is this called a "model"? The internal representation of such a time period is already there: `core/TimeDelta.*`. That would be the proper type for reading and writing such a GUI component, or not?

Am I getting somewhere?

michaelk83 commented:
I don't know the Qt details, but the idea is to make the number+unit fields reusable as a single unit. So instead of specifying `[spin box][dropdown]` in each place, you'd specify a `[time delta]`, and it would look and work the same in all such places (you could allow some minimum configuration such as allowed units, but for now, I think "days, months" or "days, weeks, months" are a good default). It can still be a `[spin box][dropdown]` internally, but you reduce the code duplication and get consistent UX.

I'll leave the more technical details to @droidmonkey et al.

schlimmchen committed:
Implement history purge feature in Database Maintenance

Fixes #7550

The database maintenance widget is complemented with a section that allows the user to clean up old history entries. the age of history entries to be removed is configurable. respective entries are searched for and removed with the click of a button. message boxes inform about whether or not and if so how many entries were removed from the database. Reuse the `maintenanceHistoryDays` attribute from KeePass2 to store the desired days of history to maintain.

schlimmchen commented:
@droidmonkey Can I ask for feedback on the progress of this PR? My feeling is that it is nearly mature. The `TimeDeltaUI` is meant to be used again for the custom expiration preset per entry feature and maybe for the "show expring entries on unlock" feature (and possibly others in the future).

![Screenshot from 2022-04-05 20-54-50](https://user-images.githubusercontent.com/3578416/161829257-4bbacf69-92ca-4832-98fd-0e82ca5d6805.png)
![Screenshot from 2022-04-05 20-56-01](https://user-images.githubusercontent.com/3578416/161829266-460fccf9-dd43-44be-9038-2e4719d0b894.png)
![Screenshot from 2022-04-05 20-55-23](https://user-images.githubusercontent.com/3578416/161829278-8d2286f7-22dc-4f1d-a5b2-7ee1b049e3c2.png)
![Screenshot from 2022-04-05 20-55-32](https://user-images.githubusercontent.com/3578416/161829283-90747256-58ec-40d6-af28-b94f8991bca5.png)

I see that `QT_TR_N_NOOP` was introduced in Qt5.12 and abviously that's not to be expected to be available... That's a major bummer, gonna have to think about that (`QT_TN_NOOP` is not enough because I do want to mark the strings as having plural forms...)

droidmonkey commented on share/translations/keepassxc_en_US.ts:
```diff
@@ -8521,6 +8521,44 @@ Kernel: %3 %4</translation>
         <translation>Weak Passwords</translation>
     </message>
 </context>
+<context>
```
This file should not be changed, it is done on release

droidmonkey commented on src/core/Metadata.cpp:
```diff
@@ -314,7 +316,9 @@ void Metadata::setDefaultUserNameChanged(const QDateTime& value)
-    set(m_data.maintenanceHistoryDays, value);
+    // KeePass2 has a hardcoded maximum of 3650 days.
+    // for compatibility we make sure not to use bigger numbers.
+    set(m_data.maintenanceHistoryDays, std::min(value, 3650));
```
should use qMin and avoid adding the include

droidmonkey commented on src/core/Entry.cpp:
```diff
@@ -755,10 +755,10 @@ void Entry::addHistoryItem(Entry* entry)
 }
 
-void Entry::removeHistoryItems(const QList<Entry*>& historyEntries)
+int Entry::removeHistoryItems(const QList<Entry*>& historyEntries)
```
This function is all sorts of strange. It accepts a random list of entries and attempts to remove them from the internal history list and... even if it doesn't own that entry in history.... it deletes it (note this is only caught when asserts are enabled). This needs to be totally rethought from an api perspective.

droidmonkey commented on src/core/TimeDelta.cpp:
```diff
@@ -82,3 +83,10 @@ int TimeDelta::getYears() const
     return m_years;
 }
+
+int TimeDelta::toDays() const
+{
+    QDateTime now = Clock::currentDateTime();
+    QDateTime future = now + *this;
+    return now.daysTo(future);
+}
```
This function is out of place in this class. It implies a comparison from current time, which none of the other calls make. This whole class seems extremely pointless to me, to be honest.

droidmonkey commented on src/core/TimeDeltaUI.h:
```diff
@@ -0,0 +1,52 @@
+class QSpinBox;
+class QComboBox;
+
+class TimeDeltaUI : public QObject
```
This class is poorly named, it is a QObject not a UI element (typically QWidget). This would be better placed as a standard time picker type widget that could also be used in the expiration UI, for example. As it is, this is a helper class that converts time units and has no need to know about UI elements like spinboxes and comboboxes. The parent widget to those elements should handle the communication.

droidmonkey commented on src/gui/dbsettings/DatabaseSettingsWidgetMaintenance.cpp:
```diff
@@ -221,3 +232,66 @@ void DatabaseSettingsWidgetMaintenance::purgeUnusedCustomIcons()
+        }
+    }
+
+    if (0 == purgeCounter) {
```
variable names come first in comparisons
